#! /bin/sh /usr/share/dpatch/dpatch-run
## 12-cgi-ignores-stderr.dpatch by  <skx@gold.my.flat>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Avoid STDERR messages from CGI scripts being sent to clients.  (Closes: #437927)

@DPATCH@
diff -urNad '--exclude=CVS' '--exclude=.svn' '--exclude=.git' '--exclude=.arch' '--exclude=.hg' '--exclude=_darcs' '--exclude=.bzr' thttpd-2.25b~/libhttpd.c thttpd-2.25b/libhttpd.c
--- thttpd-2.25b~/libhttpd.c	2005-06-29 18:50:39.000000000 +0100
+++ thttpd-2.25b/libhttpd.c	2010-03-01 18:01:43.000000000 +0000
@@ -3509,6 +3509,19 @@
     (void) signal( SIGPIPE, SIG_DFL );
 #endif /* HAVE_SIGSET */

+    /**
+     * Here we unilaterally kill output from /dev/stderr
+     */
+    {
+      int tmpfd = open("/dev/null", O_RDWR);
+
+      if ( tmpfd != -1 ) {
+        close(STDERR_FILENO);
+        dup2(tmpfd, STDERR_FILENO);
+        close(tmpfd);
+      }
+    }
+
     /* Run the program. */
     (void) execve( binary, argp, envp );

